8280682: Refactor AOT code source validation checks

calvinccheung opened the pull request:
This changset refactors CDS class paths and module paths validation code into a new class `AOTCodeSource` and related class `AOTCodeSourceConfig`. Code has been moved from filemap.[c|h]pp, classLoader.[c|h]pp, and classLoaderExt.[c|h]pp to aotCodeSource.[c|h]pp. CDS dependencies have been removed from `classLoader.cpp`. More refactoring could be done, such as removing `classLoaderExt.cpp`, in a future RFE.

Passed tiers 1 - 5 testing.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8280682](https://bugs.openjdk.org/browse/JDK-8280682): Refactor AOT code source validation checks (**Enhancement** - P4)


### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)
 * [Ashutosh Mehra](https://openjdk.org/census#asmehra) (@ashu-mehra - Committer) Review applies to [01238742](https://git.openjdk.org/jdk/pull/23476/files/01238742b1969d81528705621404c89af5234fe2)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [9e4e33dd](https://git.openjdk.org/jdk/pull/23476/files/9e4e33ddfae725083770b5fc5b839482e1571155)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [9e4e33dd](https://git.openjdk.org/jdk/pull/23476/files/9e4e33ddfae725083770b5fc5b839482e1571155)

### Contributors
 * Ioi Lam `<iklam@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23476/head:pull/23476` \
`$ git checkout pull/23476`

Update a local copy of the PR: \
`$ git checkout pull/23476` \
`$ git pull https://git.openjdk.org/jdk.git pull/23476/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23476`

View PR using the GUI difftool: \
`$ git pr show -t 23476`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23476.diff">https://git.openjdk.org/jdk/pull/23476.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23476#issuecomment-2644062222)
</details>

calvinccheung commented:
/contributor add @iklam

iklam commented on src/hotspot/share/cds/aotCodeSource.hpp:
```diff
@@ -0,0 +1,271 @@
+// during AOTCache creation is the same as when the AOTCache is used during runtime.
+// Non-existent entries are recored during AOTCache creation. Those non-existent entries
+// must not exist during runtime.
+//
```
Typos:
- "subjected to AOTCodeSourceConfig::validate()" -- the function has two parameters, but we can omit them in this comment
- "validation is performed on *the* AOTCodeSources"
- "during AOTCache creation *are* the same"
- "on-existent entries are *recorded*"

calvinccheung replied:
Fixed.

iklam commented on src/hotspot/share/cds/aotCodeSource.cpp:
```diff
@@ -0,0 +1,987 @@
+};
+
+// AllCodeSourceStreams is used to iterate over all the code sources that
+// are available to the application from -Xbootclasspath, -classpath and --module-path
```
Consider adding this comment:

// When creating an AOT cache, we store the contents from AllCodeSourceStreams
// into an array of AOTCodeSources. See AOTCodeSourceConfig::dumptime_init_helper().
// When loading the AOT cache in a production run, we compare the contents of the
// stored AOTCodeSources against the current AllCodeSourceStreams to determine whether
// the AOT cache is compatible with the current JVM. See AOTCodeSourceConfig::validate().

calvinccheung replied:
Added the comment.

iklam reviewed (commented):
Looks good. A few suggestions for the in-line comments.

calvinccheung commented on src/hotspot/share/cds/aotCodeSource.hpp:
```diff
@@ -0,0 +1,271 @@
+
+  // Functions used only during dumptime
+  static void dumptime_init(TRAPS);
+  static size_t estimate_size_for_archive() {
```
Removed. I also removed the `estimate_size_for_archive_helper()` method.

calvinccheung commented on src/hotspot/share/cds/aotCodeSource.hpp:
```diff
@@ -0,0 +1,271 @@
+  }
+
+  // Common accessors
+  int boot_start()                       const { return 1; }
```
I renamed them as follows:
```
  // Common accessors
  int boot_cp_start_index()          const { return 1; }
  int boot_cp_end_index()            const { return _boot_classpath_end; }
  int app_cp_start_index()           const { return boot_cp_end_index(); }
  int app_cp_end_index()             const { return _app_classpath_end; }
  int module_path_start_index()      const { return app_cp_end_index(); }
  int module_path_end_index()        const { return _module_end; }
```

calvinccheung commented on src/hotspot/share/cds/aotCodeSource.cpp:
```diff
@@ -0,0 +1,987 @@
+}
+
+bool AOTCodeSourceConfig::need_lcp_match_helper(int start, int end, CodeSourceStream& css) const {
+  if (app_end() == boot_start()) {
```
Fixed.

calvinccheung commented on src/hotspot/share/cds/aotCodeSource.cpp:
```diff
@@ -0,0 +1,987 @@
+// CodeSource can be matched exactly: this means all other CodeSources must be
+// matched exactly.
+bool AOTCodeSourceConfig::need_lcp_match(AllCodeSourceStreams& all_css) const {
+  if (!need_lcp_match_helper(boot_start(), boot_end(), all_css.boot_cp()) ||
```
Done.

calvinccheung commented on src/hotspot/share/cds/aotCodeSource.cpp:
```diff
@@ -0,0 +1,987 @@
+    }
+  }
+
+  if (is_boot_classpath && runtime_css.has_next() && (need_to_check_app_classpath() || num_module_paths() > 0)) {
```
I added some comment:
```
  // Check if the runtime boot classpath has more entries than the one stored in the archive and if the app classpath
  // or the module path requires validation.
  if (is_boot_classpath && runtime_css.has_next() && (need_to_check_app_classpath() || num_module_paths() > 0)) {
    // the check passes if all the extra runtime boot classpath entries are non-existent
    if (check_paths_existence(runtime_css)) {
      log_warning(cds)("boot classpath is longer than expected");
      return false;
    }
  }
```
Also fixed the `check_paths_existence()` method so it only checks the extra entries.

calvinccheung commented on src/hotspot/share/cds/filemap.cpp:
```diff
@@ -321,798 +299,56 @@ void FileMapHeader::print(outputStream* st) {
+  if (header()->has_full_module_graph() && has_extra_module_paths) {
     CDSConfig::stop_using_optimized_module_handling();
-    log_info(cds)("optimized module handling: disabled because of mismatched module paths");
+    log_info(cds)("optimized module handling: disabled because of extra module path(s) are specified");
```
Fixed.

ashu-mehra commented on src/hotspot/share/runtime/threads.cpp:
```diff
@@ -24,6 +24,7 @@
  *
  */
 
+#include "cds/aotCodeSource.hpp"
```
Why is this include needed?

calvinccheung replied:
Removed.

ashu-mehra commented:
Just one more comment, rest looks good.

dholmes-ora commented on src/hotspot/share/cds/aotCodeSource.cpp:
```diff
@@ -0,0 +1,985 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
```
If the code has been moved from other files, the current opinion/consensus is that the first copyright year should be the oldest first year from all the files from which the code was obtained.

calvinccheung replied:
Since most of the code and logic are from filemap.[c|h]pp, I've updated the copyright year to 2003, 2025.

dholmes-ora commented on src/hotspot/share/cds/aotCodeSource.hpp:
```diff
@@ -0,0 +1,268 @@
+//
+// In general, validation is performed on the AOTCodeSources to ensure the code sources used
+// during AOTCache creation are the same as when the AOTCache is used during runtime.
+// Non-existent entries are recorded during AOTCache creation. Those non-existent entries
+// must not exist during runtime.
```
Does this mean that if Foo.jar is on the classpath but does not in fact exist, then we record it was on the classpath and require it to be on the classpath at runtime, but also to still not exist?

calvinccheung replied:
Actually, we don't require the non-existent entries to be on the classpath at runtime. The appcds/NonExistClasspath.java test has cases to cover that.
So I've updated the comment as follows:
```
// Non-existent entries are recorded during AOTCache creation. Those non-existent entries,
// if they are specified at runtime, must not exist.
```
Also fixed a bug so that the behavior is the same as before this refactoring.

dholmes-ora commented on src/hotspot/share/cds/aotCodeSource.hpp:
```diff
@@ -0,0 +1,268 @@
+// must not exist during runtime.
+//
+// Some details on validation:
+// - the boot classpath could be appended during runtime if there's no app classpath and
```
```suggestion
// - the boot classpath can be appended to at runtime if there's no app classpath and no
```

calvinccheung replied:
Fixed.

dholmes-ora commented on src/hotspot/share/cds/aotCodeSource.hpp:
```diff
@@ -0,0 +1,268 @@
+// Some details on validation:
+// - the boot classpath could be appended during runtime if there's no app classpath and
+//   module path specified when an AOTCache is created;
+// - the app classpath could be appended during runtime;
```
```suggestion
// - the app classpath can be appended to at runtime;
```

calvinccheung replied:
Fixed.

dholmes-ora commented on src/hotspot/share/cds/aotCodeSource.hpp:
```diff
@@ -0,0 +1,268 @@
+// - the boot classpath could be appended during runtime if there's no app classpath and
+//   module path specified when an AOTCache is created;
+// - the app classpath could be appended during runtime;
+// - the module path during runtime could be a superset of the one specified during AOTCache creation.
```
```suggestion
// - the module path at runtime can be a superset of the one specified during AOTCache creation.
```

calvinccheung replied:
Fixed.

dholmes-ora commented on src/hotspot/share/runtime/threads.cpp:
```diff
@@ -799,15 +800,6 @@ jint Threads::create_vm(JavaVMInitArgs* args, bool* canTryAgain) {
     initialize_class(vmSymbols::jdk_internal_vm_Continuation(), CHECK_JNI_ERR);
   }
 
-#if INCLUDE_CDS
-  // capture the module path info from the ModuleEntryTable
-  ClassLoader::initialize_module_path(THREAD);
-  if (HAS_PENDING_EXCEPTION) {
-    java_lang_Throwable::print(PENDING_EXCEPTION, tty);
-    vm_exit_during_initialization("ClassLoader::initialize_module_path() failed unexpectedly");
-  }
-#endif
```
Not obvious where this functionality is now handled.

calvinccheung replied:
It is now being handled in `ClassLoaderDataShared::ensure_module_entry_tables_exist()` and `AOTCodeSourceConfig::dumptime_init_helper()`.

calvinccheung replied:
How about adding the vm_exit in `ClassLoaderDataShared::ensure_module_entry_table_exist()` instead of assert?

```
void ClassLoaderDataShared::ensure_module_entry_table_exist(oop class_loader) {
  Handle h_loader(JavaThread::current(), class_loader);
  ModuleEntryTable* met = Modules::get_module_entry_table(h_loader);
  if (met == nullptr) {
    vm_exit_during_initialization("ClassLoaderDataShared::ensure_module_entry_table_exist() failed unexpectedly");
  }
}
```

dholmes-ora replied:
I can't answer that. As a refactoring I expect to see the current behaviour preserved.

calvinccheung replied:
After discussion with Ioi, we found a place to call the vm_exit function:

```
void AOTClassLocationConfig::dumptime_init(JavaThread* current) {
  assert(CDSConfig::is_dumping_archive(), "");
  _dumptime_instance = NEW_C_HEAP_OBJ(AOTClassLocationConfig, mtClassShared);
  _dumptime_instance->dumptime_init_helper(current);
  if (current->has_pending_exception()) {
    // we can get an exception only when we run out of metaspace, but that
    // shouldn't happen this early in bootstrap.
    java_lang_Throwable::print(current->pending_exception(), tty);
    vm_exit_during_initialization("AOTClassLocationConfig::dumptime_init_helper() failed unexpectedly");
  }
}
```

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/BootClassPathMismatch.java:
```diff
@@ -235,6 +236,22 @@ public void testBootClassPathMismatchTwoJars() throws Exception {
             .assertAbnormalExit(mismatchMessage);
     }
 
+    /* Arcchive contains Hello class with only hello.jar in bootclasspath at dump time.
+     *
+     * No error - bootclasspath can be appended during runtime if no -cp is specified.
+     */
+    public void testBootClassPathAppend() throws Exception {
```
A refactoring should not be introducing new test cases. Did you refactor and enhance?

calvinccheung replied:
This is a missing testcase and works the same before and after refactoring.

calvinccheung replied:
I needed to exclude the new test from running in the dynamic archive mode because there's no -cp specified during dumping. Otherwise, the test fails with`java.lang.RuntimeException: test.dynamic.dump is not supported with an empty classpath while the classlist is not empty`.

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/NonExistClasspath.java:
```diff
@@ -67,30 +67,6 @@ static void doTest(String appJar, boolean bootcp) throws Exception {
                                  "Hello"))
             .assertNormalExit();
 
-        // Replace nonExistPath with another non-existent file in the CP, it should still work
```
Not at all clearr why these test cases have been removed?

calvinccheung replied:
I've added back the removed test cases to ensure the handling of non-existent entries remains the same after refactoring.

dholmes-ora reviewed (commented):
This is a very large change to try and digest. Is it really just a refactor? I've made a few comments in places where it seems functionality may have been changed as well.

calvinccheung commented on src/hotspot/share/cds/aotCodeSource.cpp:
```diff
@@ -0,0 +1,985 @@
+    // Arguments::get_boot_class_path() contains $JAVA_HOME/lib/modules, but we treat that separately
+    for (const char* bootcp = Arguments::get_boot_class_path(); *bootcp != '\0'; ++bootcp) {
+      if (*bootcp == *os::path_separator()) {
+        ++ bootcp;
```
Fixed.

vnkozlov commented:
Passing by comment. We touched it on recent Leyden meeting. The name "AOTCodeSource" is very confusing. Especially when we start caching AOT compiled code. Can we rename it to avoid confusion?

calvinccheung commented:
> Passing by comment. We touched it on recent Leyden meeting. The name "AOTCodeSource" is very confusing. Especially when we start caching AOT compiled code. Can we rename it to avoid confusion?

Per our discussions, I've renamed "AOTCodeSource" to "AOTClassLocation". I also renamed the related classes.

vnkozlov reviewed (approved):
Thank you for renaming classes.

dholmes-ora commented on src/hotspot/share/classfile/classLoaderDataShared.hpp:
```diff
@@ -34,7 +34,9 @@ class SerializeClosure;
 
 class ClassLoaderDataShared : AllStatic {
   static bool _full_module_graph_loaded;
+  static void ensure_module_entry_table_exist(oop class_loader);
```
```suggestion
  static void ensure_module_entry_table_exists(oop class_loader);
```

calvinccheung replied:
Fixed

dholmes-ora reviewed (approved):
A couple of minor suggestions, but otherwise nothing further from me.

Thanks

calvinccheung commented on src/hotspot/share/cds/aotClassLocation.hpp:
```diff
@@ -0,0 +1,269 @@
+// To load the AOTCache in a production run, the JVM must be using a compatible set of
+// AOTClassLocations (subjected to AOTClassLocationConfig::validate()).
+//
+// In general, validation is performed on the AOTClassLocations to ensure the code sources used
```
Fixed

calvinccheung commented on src/hotspot/share/cds/aotClassLocation.cpp:
```diff
@@ -0,0 +1,999 @@
+AOTClassLocationConfig* AOTClassLocationConfig::_dumptime_instance = nullptr;
+const AOTClassLocationConfig* AOTClassLocationConfig::_runtime_instance = nullptr;
+
+// A ClassLocationStream represents a list of code sources, which can be iterated using
```
Fixed

calvinccheung commented on src/hotspot/share/cds/aotClassLocation.cpp:
```diff
@@ -0,0 +1,999 @@
+  bool has_non_jar_modules() { return _has_non_jar_modules; }
+};
+
+// AllClassLocationStreams is used to iterate over all the code sources that
```
Fixed

calvinccheung commented on src/hotspot/share/classfile/classLoaderDataShared.cpp:
```diff
@@ -143,6 +145,21 @@ static ClassLoaderData* java_system_loader_data_or_null() {
+  ensure_module_entry_table_exist(SystemDictionary::java_system_loader());
+}
+
+void ClassLoaderDataShared::ensure_module_entry_table_exist(oop class_loader) {
```
Fixed

calvinccheung commented:
@dholmes-ora, @ashu-mehra, @vnkozlov, @iklam Thanks for the reviews!

/integrate
